### PR TITLE
API-29: Create a new attribute option

### PR DESCRIPTION
**Pull request by momoss**
Create a new attribute option

https://akeneo.atlassian.net/browse/API-29

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | -
| Added Behats                      | -
| Added integration tests           | OK
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Bundle/ApiBundle/Controller/Rest/AttributeOptionController.php [hunk 1]
```diff
@@ -25,24 +40,60 @@ class AttributeOptionController
...
     public function __construct(
         AttributeRepositoryInterface $attributeRepository,
         AttributeOptionRepositoryInterface $attributeOptionsRepository,
         NormalizerInterface $normalizer,
+        SimpleFactoryInterface $factory,
+        ObjectUpdaterInterface $updater,
+        ValidatorInterface $validator,
+        SaverInterface $saver,
+        RouterInterface $router,
+        $urlDocumentation,
```
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Controller/Rest/AttributeOptionController.php]: inject $urlDocumentation after $supportedAttributeTypes please

File: src/Pim/Bundle/ApiBundle/Controller/Rest/AttributeOptionController.php [hunk 2]
```diff
@@ -57,35 +108,173 @@ public function __construct(
      */
     public function getAction(Request $request, $attributeCode, $optionCode)
     {
+        $attribute = $this->getAttribute($attributeCode);
```
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Controller/Rest/AttributeOptionController.php]: wrong rebase ?

File: src/Pim/Bundle/ApiBundle/Controller/Rest/AttributeOptionController.php [hunk 3]
```diff
@@ -57,35 +108,173 @@ public function __construct(
...
+        }
+
+        $attributeOptionStandard = $this->normalizer->normalize($attributeOption, 'standard');
+
+        return new JsonResponse($attributeOptionStandard);
+    }
+
+    /**
+     * @param Request $request
+     * @param         $attributeCode
```
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Controller/Rest/AttributeOptionController.php]: missing type

File: src/Pim/Component/Catalog/Updater/AttributeOptionUpdater.php [hunk 4]
```diff
@@ -56,17 +58,53 @@ public function update($attributeOption, array $data, array $options = [])
 
         $isNew = $attributeOption->getId() === null;
         $readOnlyFields = ['attribute', 'code'];
-        foreach ($data as $field => $data) {
```
Comment (skeleton) [inline src/Pim/Component/Catalog/Updater/AttributeOptionUpdater.php]: :+1:

File: src/Pim/Bundle/ApiBundle/Controller/Rest/AttributeOptionController.php [hunk 5]
```diff
@@ -57,35 +108,173 @@ public function __construct(
      */
     public function getAction(Request $request, $attributeCode, $optionCode)
     {
+        $attribute = $this->getAttribute($attributeCode);
+        $this->isAttributeSupportsOptions($attribute, $attributeCode);
```
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/Controller/Rest/AttributeOptionController.php]: supporting ?

File: src/Pim/Bundle/ApiBundle/Controller/Rest/AttributeOptionController.php [hunk 6]
```diff
@@ -57,35 +108,173 @@ public function __construct(
...
+
+        $this->saver->save($attributeOption);
+
+        $response = $this->getCreateResponse($attribute, $attributeOption);
+
+        return $response;
+    }
+
+    /**
+     * Return an attribute. Throw an exception if attribute doesn't exist
```
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/Controller/Rest/AttributeOptionController.php]: a detail : could you add a dot a the end of the phpdoc sentences please

File: src/Pim/Bundle/ApiBundle/Controller/Rest/AttributeOptionController.php [hunk 7]
```diff
@@ -57,35 +108,173 @@ public function __construct(
...
         $attribute = $this->attributeRepository->findOneByIdentifier($attributeCode);
         if (null === $attribute) {
             throw new NotFoundHttpException(sprintf('Attribute "%s" does not exist.', $attributeCode));
         }
 
+        return $attribute;
+    }
+
+    /**
+     * Verify if attribute supports options
```
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/Controller/Rest/AttributeOptionController.php]: an attribute (+dot)

File: src/Pim/Bundle/ApiBundle/tests/integration/Controller/Rest/AttributeOption/CreateAttributeOptionIntegration.php [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+
+        $this->assertSame(Response::HTTP_CREATED, $response->getStatusCode());
+        $this->assertSame($attributeOptionStandard, $normalizer->normalize($attributeOption));
+    }
+
+    public function testResponseWhenContentIsInvalid()
+    {
+        $client = $this->createAuthentifiedClient();
+
+        $data =
```
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/tests/integration/Controller/Rest/AttributeOption/CreateAttributeOptionIntegration.php]: keep consistency with the other entitis by testing '{‘ please, and in a string

File: src/Pim/Bundle/ApiBundle/tests/integration/Controller/Rest/AttributeOption/CreateAttributeOptionIntegration.php [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+        $this->assertSame(Response::HTTP_BAD_REQUEST, $response->getStatusCode());
+        $this->assertSame($expectedContent, json_decode($response->getContent(), true));
+    }
+
+    public function testResponseWhenContentIsEmpty()
+    {
+        $client = $this->createAuthentifiedClient();
+
+        $data =
+<<<JSON
```
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/tests/integration/Controller/Rest/AttributeOption/CreateAttributeOptionIntegration.php]: heredoc is useless in this case, put it directly in a string

File: src/Pim/Bundle/ApiBundle/tests/integration/Controller/Rest/AttributeOption/CreateAttributeOptionIntegration.php [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+        {}
+JSON;
```
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/tests/integration/Controller/Rest/AttributeOption/CreateAttributeOptionIntegration.php]: heredoc is useless in this case, put it directly in a string

File: src/Pim/Bundle/ApiBundle/Controller/Rest/AttributeOptionController.php (hunk 2 above)
Comment (momoss) [inline src/Pim/Bundle/ApiBundle/Controller/Rest/AttributeOptionController.php]: @skeleton nop just move some code in a method

**Pull request merged**